### PR TITLE
Galaxy size

### PR DESCRIPTION
**Pull request by Lucia-Fonseca**
## Checklist
- [X] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/develop/CONTRIBUTING.md)
- [X] Write unit tests
- [X] Write documentation strings
- [X] Assign someone from your working team to review this pull request
- [X] Assign someone from the infrastructure team to review this pull request

## Suggestions
* ~The functions return the angular size in Astropy units of [1 / Mpc]. What is your opinion on that?~ The output gives only the value now.
* It does not pass all the checks because of the example in the doc-string. One of the functions is a __log-normal distribution__ and every time gives a different value. Suggestions, please!
* Suggestions for more tests, comparison with hardcore values?

### CODE REVIEW

File: skypy/galaxy/size.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    ..[1] S. Shen, H.J. Mo, S.D.M. White, M.R. Blanton, G. Kauffmann, W. Voges,
+        J. Brinkmann, I. Csabai, Mon. Not. Roy. Astron. Soc. 343, 978 (2003).
+        '''
+
+    return late_type_lognormal(magnitude, a, a, b, M0, sigma1, sigma2,
+                               size=size)
+
+
+def linear_lognormal(magnitude, a_mu, b_mu, sigma, size=None):
+    '''Lognormal distribution with linear mean.
```
Comment (philipp128) [inline skypy/galaxy/size.py:165]: Here `"""` too?

File: skypy/galaxy/size.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                     + gamma) * units.kpc
+
+    sigma_lnR = sigma2 + (sigma1 - sigma2) /\
+                         (1.0 + np.power(10, -0.8 * (magnitude - M0)))
+
+    return r_bar * np.random.lognormal(sigma=sigma_lnR, size=size)
+
+
+def early_type_lognormal(magnitude, a, b, M0, sigma1, sigma2, size=None):
+    '''Lognormal distribution for early-type galaxies.
```
Comment (philipp128) [inline skypy/galaxy/size.py:113]: `"""`?

File: skypy/galaxy/size.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    This function provides a lognormal distribution for the physical size of
+    early-type galaxies, described by equations 12, 14 and 16 in [1].
+
+    Parameters
+    ----------
+    magnitude : float or array_like.
+        Galaxy magnitude at which evaluate the lognormal distribution.
+    a_mu, b_mu : float
```
Comment (philipp128) [inline skypy/galaxy/size.py:121]: Input as `a` and `b` but documentation for `a_m` and `b_m`. Should be consistent.

File: skypy/galaxy/size.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        Linear model parameters describing the mean size of galaxies,
+        equation 3.14 in [1].
+    sigma: float
```
Comment (philipp128) [inline skypy/galaxy/size.py:124]: `sigma1, sigma2`. And documentation for `M0`?

File: skypy/galaxy/size.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        Otherwise, np.broadcast(mean, sigma).size samples are drawn.
+
+    Results
+    -------
+    physical_size : ndarray or astropy.Quantity
+        Physical distance for a given galaxy with a given magnitude, in [kpc].
+        If size is None and magnitude is a scalar, a single sample is returned.
+        If size is different from None and magnitude is scalar,
+        shape is (size,). If magnitude has shape (nm,) and size=None,
+        shape is (nm,).
```
Comment (philipp128) [inline skypy/galaxy/size.py:140]: Should this be written in doc for `size` instead of `(m,n,k)`...?

File: skypy/galaxy/size.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+def early_type_lognormal(magnitude, a, b, M0, sigma1, sigma2, size=None):
+    '''Lognormal distribution for early-type galaxies.
+    This function provides a lognormal distribution for the physical size of
+    early-type galaxies, described by equations 12, 14 and 16 in [1].
+
+    Parameters
+    ----------
+    magnitude : float or array_like.
+        Galaxy magnitude at which evaluate the lognormal distribution.
```
Comment (philipp128) [inline skypy/galaxy/size.py:120]: Maybe specify if you mean apparent or absolute magnitude.

File: skypy/galaxy/size.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        Linear model parameters describing the mean size of galaxies,
+        equation 3.14 in [1].
+    sigma: float
+        Standard deviation of the lognormal distribution for the
+        physical radius of galaxies.
+    size : int or tuple of ints, optional.
+        Output shape. If the given shape is, e.g., (m, n, k),
+        then m * n * k samples are drawn. If size is None (default),
+        a single value is returned if mean and sigma are both scalars.
+        Otherwise, np.broadcast(mean, sigma).size samples are drawn.
```
Comment (philipp128) [inline skypy/galaxy/size.py:184]: Maybe replace the description here with lines 190-193 as well?

File: skypy/galaxy/tests/test_size.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+def test_linear_lognormal():
+    """ Test lognormal distribution of galaxy sizes"""
+    # Test that a scalar input gives a scalar output
+    magnitude_scalar = -20.0
+    a_mu, b_mu, sigma = -0.24, -4.63, 0.4
+    size_scalar = size.linear_lognormal(magnitude_scalar, a_mu, b_mu, sigma)
+
+    assert np.isscalar(size_scalar.value)
+    # Test that the output has the correct units
```
Comment (philipp128) [inline skypy/galaxy/tests/test_size.py:116]: Blank line?

File: skypy/galaxy/size.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
+''' This modules computes the angular size of galaxies from
+their physical size.'''
+
+import numpy as np
+from astropy import units
+
+
+def angular_size(physical_size, redshift, cosmology):
+    '''Angular size of a galaxy.
```
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/galaxy/size.py:9]: I think the difference is very subtle. I can change it to `"""`

File: skypy/galaxy/size.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+    >>> from astropy import units
+    >>> from astropy.cosmology import FlatLambdaCDM
+    >>> cosmology = FlatLambdaCDM(H0=67.04, Om0=0.3183, Ob0=0.047745)
+    >>> size = 10.0 * units.kpc
+    >>> angular_size(size, 1, cosmology)
+    <Quantity 5.85990062e-06 rad>
+
+    References
+    ----------
+    .. [1] D. W. Hogg, (1999), astro-ph/9905116.
```
[the remark this answers is not in the document]
Comment (Lucia-Fonseca) [reply skypy/galaxy/size.py:40]: Yes, it helps CircleCI to understand this is a reference and not some array slicing.

File: skypy/galaxy/size.py (hunk 6 above)
Comment (Lucia-Fonseca) [reply skypy/galaxy/size.py:120]: good point

File: skypy/galaxy/size.py (hunk 3 above)
Comment (Lucia-Fonseca) [reply skypy/galaxy/size.py:121]: Thanks for spotting this! Copy-pasting's fault.

File: skypy/galaxy/size.py (hunk 5 above)
Comment (Lucia-Fonseca) [reply skypy/galaxy/size.py:140]: Not sure about this. This refers solely to the shape of the output.

Comment (philipp128) [reply skypy/galaxy/size.py:140]: Ja exactly. But `size` is the output shape. So it might be confusing.

**Comment by ntessore** [comment]
commenting here

**Pull request merged**